File: src/Overt.Runtime/File.cs
// Filesystem operations: File (read/write/exists/etc.), Directory
// (list/create/delete), and pure path-string helpers under Path. All
// fallible operations surface host exceptions as Result<T, IoError>.

namespace Overt.Runtime;

/// <summary>
/// File I/O companion. Mirrors the small set of file operations that Overt
/// programs need without pulling in an extern binding per call. All
/// fallible operations return <c>Result&lt;T, IoError&gt;</c>; the
/// host-side exceptions are converted to <c>IoError</c> at the boundary
/// per DESIGN.md §17. Pure path-string helpers live on <see cref="Path"/>.
/// </summary>
public static class File
{
    /// <summary>Read the file at <paramref name="path"/> as UTF-8 and
    /// return its contents. Errors (file not found, permission denied,
    /// encoding failure) surface as <c>Err(IoError)</c>.</summary>
    public static Result<string, IoError> read_to_string(string path)
    {
        try
        {
            return new ResultOk<string, IoError>(global::System.IO.File.ReadAllText(path));
        }
        catch (Exception ex)
        {
            return new ResultErr<string, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>Write <paramref name="contents"/> to <paramref name="path"/>
    /// as UTF-8, overwriting any existing file. Returns
    /// <c>Ok(())</c> on success, <c>Err(IoError)</c> on failure.</summary>
    public static Result<Unit, IoError> write_all_text(string path, string contents)
    {
        try
        {
            global::System.IO.File.WriteAllText(path, contents);
            return new ResultOk<Unit, IoError>(Unit.Value);
        }
        catch (Exception ex)
        {
            return new ResultErr<Unit, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>True iff <paramref name="path"/> names an existing file
    /// (not a directory). Predicates don't return Result; pair with
    /// <see cref="read_to_string"/> when you actually want the contents.</summary>
    public static bool exists(string path) => global::System.IO.File.Exists(path);

    /// <summary>Read the file as UTF-8, splitting on newlines. Each line
    /// excludes the trailing `\n` (and `\r\n` on Windows). The final line
    /// is included even if it lacks a trailing newline. Empty file → empty
    /// list.</summary>
    public static Result<List<string>, IoError> read_lines(string path)
    {
        try
        {
            var lines = global::System.IO.File.ReadAllLines(path);
            return new ResultOk<List<string>, IoError>(
                new List<string>(System.Collections.Immutable.ImmutableArray.Create(lines)));
        }
        catch (Exception ex)
        {
            return new ResultErr<List<string>, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>Append <paramref name="contents"/> to <paramref name="path"/>
    /// (UTF-8). Creates the file if it doesn't exist.</summary>
    public static Result<Unit, IoError> append_text(string path, string contents)
    {
        try
        {
            global::System.IO.File.AppendAllText(path, contents);
            return new ResultOk<Unit, IoError>(Unit.Value);
        }
        catch (Exception ex)
        {
            return new ResultErr<Unit, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>Delete the file at <paramref name="path"/>. Deleting a
    /// non-existent file is a no-op (matches .NET File.Delete and POSIX
    /// `rm -f`-ish semantics — programs that want a "missing" diagnostic
    /// guard with <see cref="exists"/> first).</summary>
    public static Result<Unit, IoError> delete(string path)
    {
        try
        {
            global::System.IO.File.Delete(path);
            return new ResultOk<Unit, IoError>(Unit.Value);
        }
        catch (Exception ex)
        {
            return new ResultErr<Unit, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>Size of the file in bytes. Errors (not found, permission
    /// denied, etc.) surface as Err.</summary>
    public static Result<int, IoError> size(string path)
    {
        try
        {
            var info = new global::System.IO.FileInfo(path);
            // FileInfo.Length is long; clamp to int. Files larger than
            // 2 GB are vanishingly rare for the v1 stdlib's audience and
            // can FFI to FileInfo directly when they matter.
            var len = info.Length;
            if (len > int.MaxValue)
            {
                return new ResultErr<int, IoError>(new IoError(
                    $"File.size: file '{path}' exceeds Int range ({len} bytes); use FFI for large files"));
            }
            return new ResultOk<int, IoError>((int)len);
        }
        catch (Exception ex)
        {
            return new ResultErr<int, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>Atomic-where-supported rename. On the same filesystem this
    /// is the rename(2) primitive; across filesystems .NET falls back to
    /// copy + delete. Programs that need strict-atomic semantics across
    /// filesystem boundaries handle that themselves.</summary>
    public static Result<Unit, IoError> move(string from, string to)
    {
        try
        {
            global::System.IO.File.Move(from, to);
            return new ResultOk<Unit, IoError>(Unit.Value);
        }
        catch (Exception ex)
        {
            return new ResultErr<Unit, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>Copy the file. Existing destination is overwritten —
    /// matches the conventional "cp -f" default. Programs that want a
    /// "fail if exists" check guard with <see cref="exists"/> first.</summary>
    public static Result<Unit, IoError> copy(string from, string to)
    {
        try
        {
            global::System.IO.File.Copy(from, to, overwrite: true);
            return new ResultOk<Unit, IoError>(Unit.Value);
        }
        catch (Exception ex)
        {
            return new ResultErr<Unit, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>Read the file as raw bytes. Pairs with
    /// <see cref="Bytes.from_utf8"/> / <see cref="Bytes.to_utf8"/> when
    /// callers need to round-trip text through the binary form (e.g.
    /// hashing, network framing).</summary>
    public static Result<Bytes, IoError> read_bytes(string path)
    {
        try
        {
            return new ResultOk<Bytes, IoError>(
                new Bytes(global::System.IO.File.ReadAllBytes(path)));
        }
        catch (Exception ex)
        {
            return new ResultErr<Bytes, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>Write raw bytes, overwriting any existing file. Default
    /// mode 0644 (matches WriteAllText).</summary>
    public static Result<Unit, IoError> write_bytes(string path, Bytes data)
    {
        try
        {
            global::System.IO.File.WriteAllBytes(path, data.Items);
            return new ResultOk<Unit, IoError>(Unit.Value);
        }
        catch (Exception ex)
        {
            return new ResultErr<Unit, IoError>(new IoError(ex.Message));
        }
    }
}

/// <summary>
/// Filesystem directory operations. All carry !{io}. Directory listing,
/// creation (with parents-as-needed), and removal (with optional
/// recursive flag).
/// </summary>
public static class Directory
{
    public static bool exists(string path) => global::System.IO.Directory.Exists(path);

    /// <summary>Create the directory, including any missing parents.
    /// No-op if it already exists.</summary>
    public static Result<Unit, IoError> create(string path)
    {
        try
        {
            global::System.IO.Directory.CreateDirectory(path);
            return new ResultOk<Unit, IoError>(Unit.Value);
        }
        catch (Exception ex)
        {
            return new ResultErr<Unit, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>List the entry names in the directory (file and
    /// subdirectory names; not full paths). Programs that want full
    /// paths join with <see cref="Path.join"/> per entry. The list
    /// order is filesystem-dependent and not promised stable across
    /// hosts.</summary>
    public static Result<List<string>, IoError> list(string path)
    {
        try
        {
            var entries = global::System.IO.Directory.GetFileSystemEntries(path);
            var builder = System.Collections.Immutable.ImmutableArray.CreateBuilder<string>(entries.Length);
            foreach (var e in entries)
            {
                builder.Add(global::System.IO.Path.GetFileName(e) ?? e);
            }
            return new ResultOk<List<string>, IoError>(new List<string>(builder.MoveToImmutable()));
        }
        catch (Exception ex)
        {
            return new ResultErr<List<string>, IoError>(new IoError(ex.Message));
        }
    }

    /// <summary>Delete the directory. With <paramref name="recursive"/>
    /// = true, removes all contents; with false, requires the directory
    /// to be empty (matches POSIX rmdir / rm -r split).</summary>
    public static Result<Unit, IoError> delete(string path, bool recursive)
    {
        try
        {
            global::System.IO.Directory.Delete(path, recursive);
            return new ResultOk<Unit, IoError>(Unit.Value);
        }
        catch (Exception ex)
        {
            return new ResultErr<Unit, IoError>(new IoError(ex.Message));
        }
    }
}

/// <summary>
/// Pure path-string helpers. None of these touch the filesystem — they
/// operate on the path string itself. For real file existence checks /
/// reads / writes, see <see cref="File"/>.
/// </summary>
public static class Path
{
    /// <summary>Join two path segments with the platform-appropriate
    /// separator. <c>Path.join("dir", "file.txt")</c> yields
    /// <c>"dir/file.txt"</c> on Unix or <c>"dir\\file.txt"</c> on
    /// Windows. The Go runtime does the same, so output round-trips
    /// across back ends on each platform.</summary>
    public static string join(string parent, string child)
        => global::System.IO.Path.Combine(parent, child);

    /// <summary>Directory portion of <paramref name="path"/>. Returns
    /// <c>None</c> when the path has no parent (a bare filename or
    /// the empty string).</summary>
    public static Option<string> parent(string path)
    {
        var dir = global::System.IO.Path.GetDirectoryName(path);
        if (string.IsNullOrEmpty(dir)) return new OptionNone<string>();
        return new OptionSome<string>(dir);
    }

    /// <summary>Final component of <paramref name="path"/>. Returns
    /// <c>None</c> for the empty string; otherwise the segment after
    /// the last separator.</summary>
    public static Option<string> file_name(string path)
    {
        var name = global::System.IO.Path.GetFileName(path);
        if (string.IsNullOrEmpty(name)) return new OptionNone<string>();
        return new OptionSome<string>(name);
    }

    /// <summary>File extension including the leading dot, e.g. <c>".ov"</c>.
    /// Returns <c>None</c> when the path has no extension.</summary>
    public static Option<string> extension(string path)
    {
        var ext = global::System.IO.Path.GetExtension(path);
        if (string.IsNullOrEmpty(ext)) return new OptionNone<string>();
        return new OptionSome<string>(ext);
    }

    /// <summary>Replace (or add) the extension on <paramref name="path"/>.
    /// <paramref name="ext"/> may include or omit the leading dot;
    /// empty <paramref name="ext"/> strips any existing extension.</summary>
    public static string with_extension(string path, string ext)
    {
        var stripped = global::System.IO.Path.ChangeExtension(path, null) ?? path;
        if (string.IsNullOrEmpty(ext)) return stripped;
        return ext.StartsWith('.') ? stripped + ext : stripped + "." + ext;
    }

    /// <summary>True iff <paramref name="path"/> is rooted (absolute) per
    /// the host's path conventions.</summary>
    public static bool is_absolute(string path)
        => global::System.IO.Path.IsPathRooted(path);
}

File: runtime/go/overt/file.go
package overt

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// FileReadToString reads the file at path as UTF-8 and returns its
// contents as a Result. Errors (not found, permission, encoding)
// surface as Err with the host's error message in the IoError
// narrative — same convention as the C# runtime, so a program reading
// the same path against the same file gets equivalent telemetry on
// either back end.
func FileReadToString(path string) Result[string, IoError] {
	bytes, err := os.ReadFile(path)
	if err != nil {
		return Err[string, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[string, IoError](string(bytes))
}

// FileWriteAllText writes contents to path as UTF-8, overwriting any
// existing file. Permissions are 0644 (rw-r--r--), matching the C#
// runtime's File.WriteAllText default.
func FileWriteAllText(path string, contents string) Result[Unit, IoError] {
	if err := os.WriteFile(path, []byte(contents), 0644); err != nil {
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[Unit, IoError](UnitValue)
}

// FileExists is true iff path names an existing file (not a directory).
// A directory at the path returns false; an unreadable path also
// returns false (the OS-level distinguishing of "doesn't exist" vs
// "exists but unreadable" is a sharper edge than v1 carves out).
func FileExists(path string) bool {
	info, err := os.Stat(path)
	if err != nil {
		return false
	}
	return !info.IsDir()
}

// FileReadLines reads the file as UTF-8 and splits on newlines. Each
// line excludes the trailing '\n' (and '\r\n' on Windows). The final
// line is included even without a trailing newline. Empty file →
// empty list.
func FileReadLines(path string) Result[List[string], IoError] {
	bytes, err := os.ReadFile(path)
	if err != nil {
		return Err[List[string], IoError](IoError{Narrative: err.Error()})
	}
	if len(bytes) == 0 {
		return Ok[List[string], IoError](List[string]{Items: []string{}})
	}
	// Strip a single trailing newline so the split doesn't add a
	// spurious empty trailing element. CRLF or LF.
	body := string(bytes)
	if strings.HasSuffix(body, "\r\n") {
		body = body[:len(body)-2]
	} else if strings.HasSuffix(body, "\n") {
		body = body[:len(body)-1]
	}
	// Split on '\n', then trim a trailing '\r' on each (Windows CRLF).
	parts := strings.Split(body, "\n")
	for i, p := range parts {
		if strings.HasSuffix(p, "\r") {
			parts[i] = p[:len(p)-1]
		}
	}
	return Ok[List[string], IoError](List[string]{Items: parts})
}

// FileAppendText appends contents to path (UTF-8). Creates the file
// if missing. Default mode 0644 to match FileWriteAllText.
func FileAppendText(path string, contents string) Result[Unit, IoError] {
	f, err := os.OpenFile(path, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0644)
	if err != nil {
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	defer f.Close()
	if _, err := f.WriteString(contents); err != nil {
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[Unit, IoError](UnitValue)
}

// FileDelete removes the file. Deleting a non-existent file is a no-op
// (matches .NET File.Delete and `rm -f` style — programs that want
// "missing" diagnostics guard with FileExists first).
func FileDelete(path string) Result[Unit, IoError] {
	if err := os.Remove(path); err != nil {
		// IsNotExist → silent success, matching .NET semantics.
		if os.IsNotExist(err) {
			return Ok[Unit, IoError](UnitValue)
		}
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[Unit, IoError](UnitValue)
}

// FileSize returns the file's byte count, clamped to Int range. Files
// larger than ~2 GB return Err with a guidance message — programs
// that need them FFI to os.FileInfo directly.
func FileSize(path string) Result[int, IoError] {
	info, err := os.Stat(path)
	if err != nil {
		return Err[int, IoError](IoError{Narrative: err.Error()})
	}
	if info.Size() > int64(^uint(0)>>1) /* int max */ {
		return Err[int, IoError](IoError{Narrative: fmt.Sprintf(
			"File.size: file %q exceeds Int range (%d bytes); use FFI for large files",
			path, info.Size())})
	}
	return Ok[int, IoError](int(info.Size()))
}

// FileMove renames from → to. Atomic on the same filesystem; cross-
// filesystem behavior is OS-specific (Linux: EXDEV, error; macOS / Windows:
// fall back). Programs needing strict cross-fs semantics handle the
// error themselves.
func FileMove(from string, to string) Result[Unit, IoError] {
	if err := os.Rename(from, to); err != nil {
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[Unit, IoError](UnitValue)
}

// FileReadBytes reads the file as raw bytes. Pairs with
// BytesFromUtf8 / BytesToUtf8 when round-tripping text.
func FileReadBytes(path string) Result[Bytes, IoError] {
	data, err := os.ReadFile(path)
	if err != nil {
		return Err[Bytes, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[Bytes, IoError](Bytes{Items: data})
}

// FileWriteBytes writes raw bytes, overwriting any existing file.
// Default mode 0644 to match FileWriteAllText.
func FileWriteBytes(path string, data Bytes) Result[Unit, IoError] {
	if err := os.WriteFile(path, data.Items, 0644); err != nil {
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[Unit, IoError](UnitValue)
}

// FileCopy copies src → dst. Existing dst is overwritten. Mode bits
// are not preserved beyond what the new file's umask allows; programs
// that need source-mode preservation use FFI to os.Chmod after copy.
func FileCopy(from string, to string) Result[Unit, IoError] {
	srcBytes, err := os.ReadFile(from)
	if err != nil {
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	if err := os.WriteFile(to, srcBytes, 0644); err != nil {
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[Unit, IoError](UnitValue)
}

// DirectoryExists is the directory-existence predicate (distinct from
// FileExists, which excludes directories).
func DirectoryExists(path string) bool {
	info, err := os.Stat(path)
	if err != nil {
		return false
	}
	return info.IsDir()
}

// DirectoryCreate creates the directory, including any missing
// parents. Mode 0755 (rwxr-xr-x) by default.
func DirectoryCreate(path string) Result[Unit, IoError] {
	if err := os.MkdirAll(path, 0755); err != nil {
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[Unit, IoError](UnitValue)
}

// DirectoryList returns the entry names in the directory (file and
// subdirectory names, not full paths). Order is filesystem-dependent
// and not stable across hosts.
func DirectoryList(path string) Result[List[string], IoError] {
	entries, err := os.ReadDir(path)
	if err != nil {
		return Err[List[string], IoError](IoError{Narrative: err.Error()})
	}
	out := make([]string, len(entries))
	for i, e := range entries {
		out[i] = e.Name()
	}
	return Ok[List[string], IoError](List[string]{Items: out})
}

// DirectoryDelete removes the directory. recursive=true removes all
// contents (rm -r style); false requires the directory to be empty
// (rmdir style).
func DirectoryDelete(path string, recursive bool) Result[Unit, IoError] {
	var err error
	if recursive {
		err = os.RemoveAll(path)
	} else {
		err = os.Remove(path)
	}
	if err != nil {
		return Err[Unit, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[Unit, IoError](UnitValue)
}

// PathJoin joins two path segments with the platform-appropriate
// separator. Mirrors C# Path.Combine semantics.
func PathJoin(parent string, child string) string {
	return filepath.Join(parent, child)
}

// PathParent returns the directory portion of path, or None when
// the path has no parent.
func PathParent(path string) Option[string] {
	dir := filepath.Dir(path)
	// filepath.Dir returns "." for paths with no directory, "/" for
	// the root, and the path itself for unrooted single segments.
	// Treat "." (no parent) as None to match the C# runtime's
	// "string.IsNullOrEmpty(GetDirectoryName)" behavior.
	if dir == "." || dir == "" {
		return None[string]()
	}
	return Some(dir)
}

// PathFileName returns the final segment of path, or None for the
// empty string. Mirrors filepath.Base except None instead of "."
// for paths consisting only of separators.
func PathFileName(path string) Option[string] {
	if path == "" {
		return None[string]()
	}
	name := filepath.Base(path)
	if name == "." || name == string(filepath.Separator) {
		return None[string]()
	}
	return Some(name)
}

// PathExtension returns the file extension including the leading
// dot (e.g. ".ov"), or None when the path has no extension.
func PathExtension(path string) Option[string] {
	ext := filepath.Ext(path)
	if ext == "" {
		return None[string]()
	}
	return Some(ext)
}

// PathWithExtension replaces (or adds) the extension on path. The
// supplied ext may include or omit the leading dot; both forms are
// accepted. Empty ext strips any existing extension.
func PathWithExtension(path string, ext string) string {
	stripped := strings.TrimSuffix(path, filepath.Ext(path))
	if ext == "" {
		return stripped
	}
	if !strings.HasPrefix(ext, ".") {
		ext = "." + ext
	}
	return stripped + ext
}

// PathIsAbsolute is the absolute-path predicate.
func PathIsAbsolute(path string) bool {
	return filepath.IsAbs(path)
}
